config/versions/versions.go

```diff
@@ -0,0 +1,202 @@
+			exchMethod = ExchangeVersion.DowngradeExchange
+		}
+
+		log.Printf("Running %s to config version %v\n", action, next)
+
+		patch := m.versions[next]
+
+		if cPatch, ok := patch.(ConfigVersion); ok {
+			if j, err = configMethod(cPatch, ctx, j); err != nil {
+				return j, fmt.Errorf("%w %s to %v: %w", errApplyingVersion, action, next, err)
+			}
+		}
+
+		if ePatch, ok := patch.(ExchangeVersion); ok {
+			if j, err = exchangeDeploy(ctx, ePatch, exchMethod, j); err != nil {
+				return j, fmt.Errorf("%w %s to %v: %w", errApplyingVersion, action, next, err)
+			}
+		}
+
+		current = next
+
+		if j, err = jsonparser.Set(j, []byte(strconv.Itoa(current)), "version"); err != nil {
+			return j, fmt.Errorf("%w `version` during %s to %v: %w", common.ErrSettingField, action, next, err)
+		}
+	}
+
+	log.Println("Version management finished")
+
+	return j, nil
+}
+
+func exchangeDeploy(ctx context.Context, patch ExchangeVersion, method func(ExchangeVersion, context.Context, []byte) ([]byte, error), j []byte) ([]byte, error) {
+	var errs error
+	wanted := patch.Exchanges()
+	var i int
+	eFunc := func(exchOrig []byte, _ jsonparser.ValueType, _ int, _ error) {
+		defer func() { i++ }()
+		name, err := jsonparser.GetString(exchOrig, "name")
+		if err != nil {
+			errs = common.AppendError(errs, fmt.Errorf("%w: %w `name`: %w", errModifiyingExchange, common.ErrGettingField, err))
+			return
+		}
+		for _, want := range wanted {
+			if want != "*" && want != name {
+				continue
+			}
+			exchNew, err := method(patch, ctx, exchOrig)
+			if err != nil {
+				errs = common.AppendError(errs, fmt.Errorf("%w: %w", errModifiyingExchange, err))
+				continue
+			}
+			if !bytes.Equal(exchNew, exchOrig) {
+				if j, err = jsonparser.Set(j, exchNew, "exchanges", "["+strconv.Itoa(i)+"]"); err != nil {
+					errs = common.AppendError(errs, fmt.Errorf("%w: %w `exchanges.[%d]`: %w", errModifiyingExchange, common.ErrSettingField, i, err))
+				}
+			}
+		}
+	}
+	v, dataType, _, err := jsonparser.Get(j, "exchanges")
+	switch {
+	case errors.Is(err, jsonparser.KeyPathNotFoundError), dataType != jsonparser.Array:
+		return j, nil
+	case err != nil:
+		return j, fmt.Errorf("%w: %w `exchanges`: %w", errModifiyingExchange, common.ErrGettingField, err)
+	}
+	if _, err := jsonparser.ArrayEach(bytes.Clone(v), eFunc); err != nil {
+		return j, err
+	}
+	return j, errs
+}
+
+// registerVersion takes instances of config versions and adds them to the registry
+// Versions should be added sequentially without gaps, in import.go init
+// Any errors will also added to the registry for reporting later
+func (m *manager) registerVersion(v any) {
+	m.m.Lock()
+	defer m.m.Unlock()
+	ver, err := strconv.Atoi(strings.TrimSuffix(strings.TrimPrefix(fmt.Sprintf("%T", v), "*v"), ".Version"))
+	if err != nil {
+		m.errors = common.AppendError(m.errors, fmt.Errorf("%w '%T': %w", errRegisteringVersion, v, err))
+		return
+	}
+	switch v.(type) {
+	case ExchangeVersion, ConfigVersion:
```

Comment:
Any issues if v conforms to both at the same time? 🤯

Reply:
Nope. A version can be both.

Fixed 727d1407f 📖